lib/puppet/functions/stdlib/fqdn_rotate.rb

```diff
@@ -0,0 +1,66 @@
+# frozen_string_literal: true
+
+# @summary Rotates an array or string a random number of times, combining the `fqdn` fact and an optional seed for repeatable randomness.
+Puppet::Functions.create_function(:'stdlib::fqdn_rotate') do
+  # @param input
+  #   The String you want rotated a random number of times
+  # @param seeds
+  #   One of more values to use as a custom seed. These will be combined with the host's FQDN
+  #
+  # @return [String] Returns the rotated String
+  #
+  # @example Rotating a String
+  #   stdlib::fqdn_rotate('abcd')
+  # @example Using a custom seed
+  #   stdlib::fqdn_rotate('abcd', 'custom seed')
+  dispatch :fqdn_rotate_string do
+    param 'String', :input
+    optional_repeated_param 'Variant[Integer,String]', :seeds
+    return_type 'String'
+  end
+
+  # @param input
+  #   The Array you want rotated a random number of times
+  # @param seeds
+  #   One of more values to use as a custom seed. These will be combined with the host's FQDN
+  #
+  # @return [String] Returns the rotated Array
+  #
+  # @example Rotating an Array
+  #   stdlib::fqdn_rotate(['a', 'b', 'c', 'd'])
+  # @example Using custom seeds
+  #   stdlib::fqdn_rotate([1, 2, 3], 'custom', 'seed', 1)
+  dispatch :fqdn_rotate_array do
+    param 'Array', :input
+    optional_repeated_param 'Variant[Integer,String]', :seeds
+    return_type 'Array'
+  end
+
+  def fqdn_rotate_array(input, *seeds)
+    # Check whether it makes sense to rotate ...
+    return input if input.size <= 1
+
+    result = input.clone
+
+    require 'digest/md5'
+    seed = Digest::MD5.hexdigest([fqdn_fact, seeds].join(':')).hex
```

Comment:
sa  #1342 ][
```suggestion
    require 'digest/sha2'
    seed = Digest::SHA256.hexdigest([fqdn_fact, seeds].join(':')).hex
```

Reply:
I did look at this, but it would be a breaking change...

I don't think anybody using this wants their _deterministic_ rotated things changed just because we think SHA256 would create a better _seed_ for https://github.com/puppetlabs/puppet/blob/f431150d36545626af9ad8602cdd6a7461ca025b/lib/puppet/util.rb#L725

Unwanted changes can mean services restarting, refresh_only execs being trigger etc.

The only point in the `Digest::MD5.hexdigest('seed').hex` was to turn `seed` into a repeatable integer (as `Random.new` only accepts integers for seeds)

And then there's no real point that this seed is then fed to `Random.new`.  All we are trying to do is generate a repeatable number of times, (`offset.times`), to rotate the input characters (or array) by.

Honestly I think `(seed % result.size).times` would have been good enough.

Reply:
ack makes senses and agree this is minor and doesn't warrant the distruption

Reply:
It may matter on FIPS systems where md5 may not be available, but that's not the point of this PR.